src/Sentry.Extensions.Logging/SentryEFCoreListener.cs

```diff
@@ -0,0 +1,156 @@
+            QueryExecution,
+            QueryCompiler
+        }
+
+        internal const string EFConnectionOpening = "Microsoft.EntityFrameworkCore.Database.Connection.ConnectionOpening";
+        internal const string EFConnectionClosed = "Microsoft.EntityFrameworkCore.Database.Connection.ConnectionClosed";
+        internal const string EFCommandExecuting = "Microsoft.EntityFrameworkCore.Database.Command.CommandExecuting";
+        internal const string EFCommandExecuted = "Microsoft.EntityFrameworkCore.Database.Command.CommandExecuted";
+        internal const string EFCommandFailed = "Microsoft.EntityFrameworkCore.Database.Command.CommandError";
+
+        /// <summary>
+        /// Used for EF Core 2.X and 3.X. 
+        /// <seealso href="https://docs.microsoft.com/dotnet/api/microsoft.entityframeworkcore.diagnostics.coreeventid.querymodelcompiling?view=efcore-3.1"></seealso>
+        /// </summary>
+        internal const string EFQueryStartCompiling = "Microsoft.EntityFrameworkCore.Query.QueryCompilationStarting";
+        /// <summary>
+        /// Used for EF Core 2.X and 3.X.
+        /// <seealso href="https://docs.microsoft.com/dotnet/api/microsoft.entityframeworkcore.diagnostics.coreeventid.querymodelcompiling?view=efcore-3.1"></seealso>
+        /// </summary>
+        internal const string EFQueryExecutionPlanned = "Microsoft.EntityFrameworkCore.Query.QueryExecutionPlanned";
+        internal const string EFQueryCompiling = "Microsoft.EntityFrameworkCore.Query.QueryModelCompiling";
+        internal const string EFQueryCompiled = "Microsoft.EntityFrameworkCore.Query.QueryExecutionPlanned";
+
+        private IHub _hub { get; }
+        private SentryOptions _options { get; }
+
+        private AsyncLocal<WeakReference<ISpan>> _spansCompilerLocal = new();
+        private AsyncLocal<WeakReference<ISpan>> _spansQueryLocal = new();
+        private AsyncLocal<WeakReference<ISpan>> _spansConnectionLocal = new();
+
+        public SentryEFCoreListener(IHub hub, SentryOptions options)
+        {
+            _hub = hub;
+            _options = options;
+        }
+
+        private ISpan? AddSpan(SentryEFSpanType type, string operation, string? description)
+        {
+            if (_hub.GetSpan()?.StartChild(operation, description) is { } span)
+            {
+                if (type switch
+                {
+                    SentryEFSpanType.QueryCompiler => _spansCompilerLocal,
+                    SentryEFSpanType.QueryExecution => _spansQueryLocal,
+                    SentryEFSpanType.Connection => _spansConnectionLocal,
+                    _ => null
+                } is { } asyncLocalSpan)
+                {
+                    asyncLocalSpan.Value = new WeakReference<ISpan>(span);
+                }
+                return span;
+            }
+            return null;
+        }
+
+        private ISpan? TakeSpan(SentryEFSpanType type)
+        {
+            if (type switch
+            {
```

Comment:
The `{` is of the `switch` and not the `if` so it's rather confusing with this indentation